src/common/snippets/src/pass/explicit_transpose_matmul_inputs.cpp

```diff
@@ -2,79 +2,98 @@
-#include "snippets/pass/transpose_decomposition.hpp"
+
 #include "snippets/op/subgraph.hpp"
+#include "snippets/itt.hpp"
 
-#include "openvino/core/rt_info.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
 #include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/core/rt_info.hpp"
+
+
+void ov::snippets::pass::ExplicitTransposeMatMulInputs::extract(const ov::Input<ov::Node>& input) {
+    auto parent = input.get_source_output().get_node_shared_ptr();
+    auto transpose = ngraph::as_type_ptr<ov::op::v1::Transpose>(parent);
+    while (!transpose && !ov::is_type<ov::op::v0::Parameter>(parent)) {
+        // We can set supported order and transposed_<a|b>=false only if ops have scalar shapes to avoid shape mismatching
+        const auto parent_count = parent->inputs().size();
+        bool are_weights_scalar = true;
+        for (size_t j = 1; j < parent_count; ++j) {
+            are_weights_scalar = are_weights_scalar && ngraph::shape_size(parent->get_input_shape(j)) == 1;
+        }
+        if (!are_weights_scalar)
+            break;
+
+        parent = parent->get_input_node_shared_ptr(0);
+        transpose = ngraph::as_type_ptr<ov::op::v1::Transpose>(parent);
+    }
+
+    // If there isn't another Transpose, need to create new Transpose
+    if (transpose) {
+        const auto transpose_pattern = ngraph::as_type_ptr<ov::op::v0::Constant>(transpose->get_input_node_shared_ptr(1));
+        OPENVINO_ASSERT(transpose_pattern,
+                        "ExplicitTransposeMatMulInputs expects existing Transpose with Constant order");
+
+        auto transposed_order = transpose_pattern->cast_vector<int32_t>();
+        OPENVINO_ASSERT(transposed_order.size() > 2, "Incorrect Transpose order for ExplicitTransposeMatMulInputs");
+        std::swap(*transposed_order.rbegin(), *(transposed_order.rbegin() + 1));
+
+        auto new_transpose_order = std::make_shared<ov::op::v0::Constant>(transpose_pattern->get_element_type(),
+                                                                          ngraph::Shape{transposed_order.size()},
+                                                                          transposed_order);
+        new_transpose_order->set_friendly_name(transpose_pattern->get_friendly_name());
+        ov::copy_runtime_info(transpose_pattern, new_transpose_order);
+        transpose->set_argument(1, new_transpose_order);
+        return;
+    }
+
+    // Create new Transpose before Parameter
+    OPENVINO_ASSERT(ov::is_type<opset1::Parameter>(parent),
+                    "ExplicitTransposeMatMulInputs expects Parameter in cases when there isn't existing Transpose on input");
+    const auto& consumers = parent->get_output_target_inputs(0);
+    OPENVINO_ASSERT(consumers.size() == 1,
+                    "ExplicitTransposeMatMulInputs expects Parameter with one consumer in cases when there isn't existing Transpose on input");
+    // Extract Transpose from MatMul
+    const auto rank = input.get_shape().size();
```

Comment:
Please double-check that shape is static in order to use `get_shape()`

Reply:
Done, thank you for the review!